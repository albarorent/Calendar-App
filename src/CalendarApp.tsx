import { Toaster } from "sonner";
import { AppRouter } from "./router";

export const CalendarApp = () => {
  return (
    <>
      <AppRouter />
      <Toaster />
    </>
  );
};
