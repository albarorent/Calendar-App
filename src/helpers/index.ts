export * from './calendarLocalizer'
export * from './getEnvVariables'
export * from './convertEventsToDayEvents'
export * from './getMessages'