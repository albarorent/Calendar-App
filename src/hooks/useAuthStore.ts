import { useDispatch } from "react-redux";
import { useAppSelector } from "../store/hooks";
import { calendarApi } from "../api";
import { clearErrorMessage, onChecking, onLogin, onLogout } from "../store";
import { GetResponseLogin } from "../interfaces/AuthInterface";
import { isAxiosError } from "axios";

export const useAuthStore = () => {
  const { status, user, errorMessage } = useAppSelector((state) => state.auth);
  const dispatch = useDispatch();

  const startLogin = async (email: string, password: string) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post<GetResponseLogin>("/auth/login", {
        email,
        password,
      });
      localStorage.setItem("token", data.token);
      localStorage.setItem("token-init-date", new Date().getTime().toString());
      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch (error) {
      if (isAxiosError(error)) {
        dispatch(
          onLogout(error.response?.data?.msg || "Credenciales incorrectas")
        );
      }
      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 10);
    }
  };

  const startRegister = async (
    email: string,
    password: string,
    name: string
  ) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post<GetResponseLogin>(
        "/auth/register",
        {
          email,
          password,
          name,
        }
      );
      localStorage.setItem("token", data.token);
      localStorage.setItem("token-init-date", new Date().getTime().toString());
      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch (error) {
      if (isAxiosError(error)) {
        dispatch(onLogout(error.response?.data?.msg || "Error en el registro"));
      }
      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 10);
    }
  };

  const checkAuthToken = async () => {
    const token = localStorage.getItem("token");
    if (!token) {
      dispatch(onLogout("Token no encontrado"));
      return;
    }
    try {
      const { data } = await calendarApi.get<GetResponseLogin>(
        "/auth/autenticated"
      );
      localStorage.setItem("token", data.token);
      localStorage.setItem("token-init-date", new Date().getTime().toString());
      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch (error) {
      if (isAxiosError(error)) {
        dispatch(
          onLogout(error.response?.data?.msg || "Error al validar token")
        );
      }
      localStorage.clear();
      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 10);
    }
  };

  const startLogout = () => {
    localStorage.clear();
    dispatch(onLogout("Se ha cerrado la sesi√≥n"));
  };

  return {
    status,
    user,
    errorMessage,
    startLogin,
    startRegister,
    checkAuthToken,
    startLogout,
  };
};
