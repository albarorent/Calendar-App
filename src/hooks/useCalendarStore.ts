import { EventCalendar } from "../interfaces/CalendarInterface";
import {
  onAddNewEvent,
  onDeleteEvent,
  onSetActiveEvent,
  onUpdateEvent,
} from "../store";
import { useAppDispatch, useAppSelector } from "../store/hooks";

export const useCalendarStore = () => {
  const { events, activeEvent } = useAppSelector((state) => state.calendar);

  const dispatch = useAppDispatch();
  const setActiveEvent = (calendarEvent: EventCalendar) => {
    dispatch(onSetActiveEvent(calendarEvent));
  };

  const startSavingEvent = async (calendarEvent: EventCalendar) => {
    if (calendarEvent._id) {
      //Actualizando
      dispatch(onUpdateEvent({ ...calendarEvent }));
    } else {
      dispatch(onAddNewEvent({ ...calendarEvent, _id: new Date().getTime() }));
    }
  };

  const startDeletingEvent = () => {
    dispatch(onDeleteEvent());
  };

  return {
    //* Propiedades
    events,
    activeEvent,
    hasEventSelected: !!activeEvent,
    //* MÃ©todos
    setActiveEvent,
    startSavingEvent,
    startDeletingEvent,
  };
};
