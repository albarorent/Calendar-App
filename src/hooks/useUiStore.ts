import { onCloseDateModal, onOpenDateModal } from "../store";
import { useAppDispatch, useAppSelector } from "../store/hooks";

export const useUiStore = () => {
  const dispatch = useAppDispatch();

  const { isDateModalOpen } = useAppSelector((state) => state.ui);

  const openDateModal = () => {
    dispatch(onOpenDateModal());
  };

  const closeDateModal = () => {
    dispatch(onCloseDateModal());
  };

  const toggleDateModal = () => {
    isDateModalOpen ? openDateModal() : closeDateModal();
  };

  return {
    isDateModalOpen,
    openDateModal,
    closeDateModal,
    toggleDateModal
  };
};
