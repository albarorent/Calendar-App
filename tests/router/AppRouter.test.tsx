import { AppRouter } from "../../src/router/AppRouter";
import React from "react";
import { render, screen } from "@testing-library/react";
import { useAuthStore } from "../../src/hooks";
import { MemoryRouter } from "react-router-dom";
jest.mock("../../src/hooks/useAuthStore");

const mockCheckAuthToken = jest.fn();

describe("Pruebas en <AppRouter />", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("debe de mostrar la pantalla de carga y llamar el checkAuthToken", () => {
    (useAuthStore as jest.Mock).mockReturnValue({
      status: "checking",
      checkAuthToken: mockCheckAuthToken,
    });
    render(<AppRouter />);
    expect(screen.getByText("Cargando...")).toBeTruthy();
    expect(mockCheckAuthToken).toHaveBeenCalled();
  });

  test("debe de mostrar el login si no estÃ¡ autenticado", () => {

    (useAuthStore as jest.Mock).mockReturnValue({
      status: "not-authenticated",
      checkAuthToken: mockCheckAuthToken,
    });
    const { container } = render(<MemoryRouter><AppRouter /></MemoryRouter>);
    expect(screen.getByText("Ingreso")).toBeTruthy();
    expect(container).toMatchSnapshot();
  }
  );
});
