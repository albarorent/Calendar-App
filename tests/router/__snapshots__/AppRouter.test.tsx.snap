// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AppRouter /> debe de mostrar el login si no est치 autenticado 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Ingreso
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="email"
              placeholder="Correo"
              type="text"
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="password"
              placeholder="Contrase침a"
              type="password"
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Registro
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="name"
              placeholder="Nombre"
              type="text"
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="email"
              placeholder="Correo"
              type="email"
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="password"
              placeholder="Contrase침a"
              type="password"
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="password2"
              placeholder="Repita la contrase침a"
              type="password"
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Crear cuenta"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
